ulm-semantics/test/execution.md

```diff
@@ -63,68 +69,71 @@ module ULM-TEST-EXECUTION
-    // for mocking tests
-    rule <k> UTH:ULMTestTypeHolder ~> EI:ExecutionItem => EI ~> UTH ... </k>
-    rule <k> UTHL:ULMTestTypeHolderList ~> EI:ExecutionItem => EI ~> UTHL ... </k>
-    rule <k> UTH1:ULMTestTypeHolder ~> UTH2:ULMTestTypeHolder
-                => (UTH1, UTH2):ULMTestTypeHolderList ... </k>
-    rule <k> UTH:ULMTestTypeHolder ~> UTHL:ULMTestTypeHolderList
-                => (UTH, UTHL):ULMTestTypeHolderList ... </k>
-
-    rule <k> hold_string_from_test_stack => ptr_holder P ... </k>
-         <test-stack> ListItem(P) L:List => L </test-stack>
-    rule <k> ptr_holder ptrValue(_, V) => value_holder V ... </k>
-
-
-    // TODO: Rework the implementation of the productions related to list value holding
-    // Ref - https://github.com/Pi-Squared-Inc/rust-demo-semantics/pull/167#discussion_r1813386536
-    rule <k> hold_list_values_from_test_stack => list_ptrs_holder L ~> list_values_holder .List ... </k>
-         <test-stack> L:List => .List </test-stack>
-    rule <k> list_ptrs_holder ListItem(I) LPH ~> list_values_holder LLH
-                => I ~> list_ptrs_holder LPH ~> list_values_holder LLH ... </k>
-    rule <k> ptrValue(_, V) ~> list_ptrs_holder LPH ~> list_values_holder LLH
-                => list_ptrs_holder LPH ~> list_values_holder ListItem(V) LLH ... </k>
-    rule <k> list_ptrs_holder .List => .K ... </k>
-
-    rule <k> hold I => value_holder I ... </k>
-
-    rule <k> encode_call_data_to_string
-             ~> list_values_holder ARGS , list_values_holder PTYPES , value_holder FNAME , .ULMTestTypeHolderList
-             => Bytes2String(encodeCallData(FNAME, PTYPES, ARGS))
-            ...
-         </k>
-
-    rule <k> encode_call_data_to_string
-             ~> value_holder FNAME
-             => Bytes2String(encodeCallData(FNAME, .List, .List))
-            ...
-         </k> [owise]
+    syntax BytesOrError ::= extractCallSignature(EncodeCall)  [function, total]
+    rule extractCallSignature(Fn:Identifier ( Args:EncodeArgs ))
+        => methodSignature(IdentifierToString(Fn), encodeArgsToNormalizedParams(Args))
+    syntax NormalizedFunctionParameterList ::= encodeArgsToNormalizedParams(EncodeArgs)  [function, total]
+    rule encodeArgsToNormalizedParams(.EncodeArgs) => .NormalizedFunctionParameterList
+    rule encodeArgsToNormalizedParams(_:Expression : T:Type ,  Eas:EncodeArgs )
+        => #token("#unused", "Identifier"):T , encodeArgsToNormalizedParams(Eas)
+
+    syntax Identifier ::= "append_bytes_raw"  [token]
+                        | "buffer_id"  [token]
+                        | "bytes_hooks"  [token]
+                        | "empty"  [token]
+                        | "str"  [token]
+
+    syntax EncodeValues ::= encodeArgsToEncodeValues(EncodeArgs)  [function, total]
+    rule encodeArgsToEncodeValues(.EncodeArgs) => .EncodeValues
+    rule encodeArgsToEncodeValues(E:Expression : T:Type , Es:EncodeArgs)
+        => E : T , encodeArgsToEncodeValues(Es)
```

Comment:
It looks like `EncodeArg` and `EncodeValue` have the same syntax. Is this the intended behavior? At the moment I don't fully understand how these sorts are being used.

Reply:
They are mostly the same thing. There are some obscure reasons for keeping them separate (mostly I was hesitant about using `EncodeValues` for parsing), but I think that you're probably right and I should merge them. See the last commit of this PR. If you decide that, for some reasons, they should be separate, I can create a new PR undoing that commit.